Implement global `deep_equals` function

Geometror opened the pull request:
![grafik](https://github.com/user-attachments/assets/7bcbf833-5609-495b-9f0f-738d3e0f1ccc)

A deep equality check for all variant types is something I have long missed, so this PR implements it - and I'm sure I'm not the only one. (I needed to implement this for most of my projects)
Since this is an essential feature, addressing a need discussed several times in forums etc., but never formally tracked, I allowed myself to directly implement this without a proposal. Well, and while it works in GDScript, it's much slower than a native solution. In addition, implementing this yourself can be error-prone, so the goal of this PR is to provide a solid implementation that behaves expectedly.
But if a formal proposal is desired, I could create one.

Includes unit tests.

# How it works

- Compares basic types using `==` 
- Compares `Object`, `Dictionary` and `Array` recursively
	- Compares arrays element wise (like `==`)
	- Compares dictionaries order-independently
	- Compares objects by their properties (which includes metadata) order-independently

# Why another function?

## `is_same`
"Returns `true`, for value types, if `a` and `b` share the same value. Returns `true`, for reference types, if the references of `a` and `b` are the same." (quoted from docs)
(`is_same` uses the `==` operator for `Object`s, so they are compared by object ID)

## `==` and `recursive_equal`
- the `==` operator uses `Variant::hash_compare` which calls `recursive_equal` (only exposed for `Dictionary`) on `Array` and `Dictionary`, whereby `Object`s are compared by reference (object ID)
- Changing the behavior of `==` for `Object`s would be a massive compat-breakage (so we can't just modify `Variant::hash_compare`)

There is no way to do a predictable deep equality check when using the hash function (since for object two content-equal object have different hashes), so I went with a predictable and solid deep equality check

# Remarks & Discussion

~~- Should it be named something like `deep_equals` to make its purpose clearer?~~ Yes :)

passivestar commented:
> Should it be named something like deep_equals to make its purpose clearer?

Yeah. I always assume shallow comparison unless explicitly specified otherwise

This PR doesn't account for cyclic dependencies so I think this will crash:

```gdscript
var a = {"a": 1, "b": 2}
var b = {"a": 1, "b": 2}
a["ref"] = b
b["ref"] = a
print(is_equal(a, b))
```

Cretezy commented on core/variant/variant_utility.cpp:
```diff
@@ -1234,6 +1234,108 @@ bool VariantUtilityFunctions::is_same(const Variant &p_a, const Variant &p_b) {
+	const Variant::Type type_a = p_a.get_type();
+	const Variant::Type type_b = p_b.get_type();
+
+	if (type_a != type_b) {
```
Should `1` (integer) and `1.0` (float) be equal?

Geometror replied:
No, I don't think so. These are different numerical types.

AThousandShips commented:
Should this just use `Variant::identity_compare` or `Variant::hash_compare`?

Mickeon commented:
It should probably be called `deep_equals` to make its purpose much clearer. Otherwise at a glance it looks like a more complex `==` operator. A bit iffy on naming it after "equality" too, because it's actually checking for "equivalence", but Godot is also no stranger to mixing the terms up. Unsure...

Mickeon commented:
Also, most notably, there's a few similar methods. Adding one more may make things more confusing, due of the small intricacies of each one. But again, it's a few.
Namely, Dictionary already has `recursive_equal`. Although, one may consider deprecating it in favour of the all-encompassing function this PR provides? Or just letting both coexist and share the same behaviour, or making the difference clearer.
the global scope's `is_same` comes to mind, as well.

Ivorforce commented:
> > Should it be named something like deep_equals to make its purpose clearer?
> 
> 
> 
> Yeah. I always assume shallow comparison unless explicitly specified otherwise
> 
> 
> 
> This PR doesn't account for cyclic dependencies so I think this will crash:
> 
> 
> 
> ```gdscript
> 
> var a = {"a": 1, "b": 2}
> 
> var b = {"a": 1, "b": 2}
> 
> a["ref"] = b
> 
> b["ref"] = a
> 
> print(is_equal(a, b))
> 
> ```

This could be solved with https://github.com/godotengine/godot-proposals/issues/11919 / see linked issues.

Geometror commented:
Regarding the cyclic references problem: I would go for the simplest solution in this case and use a maximum comparison depth of 128, in addition to maybe printing a warning.
Of course there are more sophisticated solutions, however, you pay the cost for those even when comparing non-cyclic stuff, which is the common case I'd argue.

Geometror committed:
Implement global deep_equals function

Geometror commented:
Renamed to `deep_equals` and added the recursion depth limit.

Just for reference:
## `is_same`
"Returns `true`, for value types, if `a` and `b` share the same value. Returns `true`, for reference types, if the references of `a` and `b` are the same." (quoted from docs)
(`is_same` uses the `==` operator for `Object`s, so they are compared by object ID)

## `==` and `recursive_equal`
- the `==` operator uses `Variant::hash_compare` which calls `recursive_equal` (only exposed for `Dictionary`) on `Array` and `Dictionary`, whereby `Object`s are compared by reference (object ID)
- Changing the behavior of `==` for `Object`s would be a massive compat-breakage (so we can't just modify `Variant::hash_compare`)

Geometror commented:
@AThousandShips 
> Should this just use Variant::identity_compare or Variant::hash_compare, outside of Object?

Variant types other than `Array`, `Dictionary` and `Object` use `==` which calls `Variant::hash_compare`. However, using `==`/`Variant::hash_compare` also for arrays and dictionaries breaks when they contain objects.
